app/gui2/src/components/DocumentationPanel.vue

```diff
@@ -0,0 +1,266 @@
+    if ('Function' in doc) {
+      return [doc.Function]
+    } else {
+      console.error(`Unexpected docs type: ${docs}, expected Function`)
+      return []
+    }
+  })
+}
+
+function asTypeDocs(docs: Docs[]): TypeDocs[] {
+  return docs.flatMap((doc) => {
+    if ('Type' in doc) {
+      return [doc.Type]
+    } else {
+      console.error(`Unexpected docs type: ${docs}, expected Type`)
+      return []
+    }
+  })
+}
+
+type DocsHandle = (db: SuggestionDb, entry: SuggestionEntry) => Docs
+
+const handleFunction: DocsHandle = (db, entry) => ({
+  Function: {
+    name: entry.name,
+    arguments: entry.arguments,
+    sections: filterSections(entry.documentation),
+  },
+})
+
+const handleDocumentation: Record<SuggestionKind, DocsHandle> = {
+  [SuggestionKind.Function]: handleFunction,
+  [SuggestionKind.Method]: handleFunction,
+  [SuggestionKind.Constructor]: handleFunction,
+  [SuggestionKind.Local]: (db, entry) => ({
+    Local: {
+      name: entry.name,
+      sections: filterSections(entry.documentation),
+    },
+  }),
+  [SuggestionKind.Type]: (db, entry) => {
+    const [entryId] = db.nameToId.lookup(entryQn(entry))
+    if (!entryId) return { Placeholder: 'No documentation available' }
+    return {
+      Type: {
+        name: entry.name,
+        arguments: entry.arguments,
+        sections: filterSections(entry.documentation),
+        methods: asFunctionDocs(getChildren(db, entryId, SuggestionKind.Method)),
+        constructors: asFunctionDocs(getChildren(db, entryId, SuggestionKind.Constructor)),
+      },
+    }
+  },
+  [SuggestionKind.Module]: (db, entry) => {
+    const [entryId] = db.nameToId.lookup(entryQn(entry))
+    if (!entryId) return { Placeholder: 'No documentation available' }
+    return {
+      Module: {
+        name: entry.name,
+        sections: filterSections(entry.documentation),
+        types: asTypeDocs(getChildren(db, entryId, SuggestionKind.Type)),
+        methods: asFunctionDocs(getChildren(db, entryId, SuggestionKind.Method)),
+      },
+    }
+  },
+}
+
+function lookupDocumentation(db: SuggestionDb, entry: SuggestionEntry): Docs {
+  const handle = handleDocumentation[entry.kind]
+  return handle ? handle(db, entry) : { Placeholder: `Entry kind not handled: ${entry.kind}` }
+}
+
+// === Helper variables ===
+
+const documentation = computed<Docs>(() => {
+  const db = useSuggestionDbStore()
+  const entry = props.selectedEntry
+  return entry ? lookupDocumentation(db.entries, entry) : { Placeholder: 'No entry selecter' }
+})
+
+const sections = computed<Sections>(() => {
+  const docs: Docs = documentation.value
+  const fallback = { tags: [], synopsis: [], examples: [] }
+  if ('Function' in docs) {
+    return docs.Function.sections
+  } else if ('Type' in docs) {
+    return docs.Type.sections
+  } else if ('Module' in docs) {
+    return docs.Module.sections
+  } else if ('Local' in docs) {
+    return docs.Local.sections
+  } else {
+    return fallback
+  }
+})
```

Comment:
Is there some way to make it less painful with Typescript? In pure JS, `docs.Function.sections || docs.Type.sections || … || fallback` works, but Typescript does not like it. It has a point I guess, but it still sucks.

Reply:
no - typescript doesn't like this kind of discriminated union in general - it only has dedicated support for unions discriminated by the value of a specific property: [more info](https://basarat.gitbook.io/typescript/type-system/discriminated-unions).
i believe the only way to make this work, would be to add `Function?: undefined, Type?: undefined` etc. to every variant.

iirc this is something i brought up when pawel introduced the initial AST API, but i guess the conclusion was, it's not worth risking breaking the existing API (and/or making the API inconsistent between languages) for slightly better ergonomics.